engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala

```diff
@@ -184,20 +215,75 @@ class WorkflowActor(workflowId: WorkflowId,
   when(WorkflowFailedState) { FSM.NullFunction }
   when(WorkflowSucceededState) { FSM.NullFunction }
 
-  onTransition {
-    case oldState -> terminalState if terminalState.terminal =>
-      log.info(s"$tag transition from $oldState to $terminalState: shutting down")
-      context.stop(self)
-    case fromState -> toState =>
-      log.info(s"$tag transitioning from $fromState to $toState")
-  }
-
   whenUnhandled {
+    case Event(MetadataServiceFailureResponse(error, action), _) =>
+      // Do something useful here??
+      log.warning(s"$tag Received MetadataServiceFailure")
+      stay
     case Event(AbortWorkflowCommand, WorkflowActorData(Some(actor), _)) =>
       actor ! EngineLifecycleActorAbortCommand
       goto(AbortingWorkflowState)
     case unhandledMessage =>
       log.warning(s"$tag received an unhandled message $unhandledMessage in state $stateName")
       stay
   }
+
+  onTransition {
+    case fromState -> toState =>
+      log.info(s"$tag transitioning from $fromState to $toState")
+      // This updates the workflow status
+      pushCurrentStateToMetadataService(toState.workflowState) onFailure {
+        case reason: Throwable => self ! MetadataServiceFailureResponse(reason)
+      }
+  }
+
+  onTransition {
+    case oldState -> terminalState if terminalState.terminal =>
+      log.info(s"$tag transition from $oldState to $terminalState: shutting down")
+      // Add the end time of the workflow in the MetadataService
+      val metadataEventMsg = MetadataEvent(MetadataKey(workflowId, None, WorkflowMetadataKeys.EndTime), MetadataValue(DateTime.now.toString))
+      serviceRegistryActor.publishToMetadataService(metadataEventMsg) map handleMetadataServiceResponse recover {
+        case reason: Throwable =>
+          self ! MetadataServiceFailureResponse(reason)
+      } andThen {
+        case _ =>
+          terminalState match {
+            case WorkflowSucceededState =>
+              context.parent ! WorkflowSucceededResponse(workflowId)
+            case WorkflowFailedState =>
+              context.parent ! WorkflowFailedResponse(workflowId, oldState, Seq.empty[Throwable])
+            case WorkflowAbortedState =>
+              context.parent ! WorkflowAbortedResponse(workflowId)
+            case unknownState => log.warning(s"tag $unknownState is an unhandled terminal state!")
+          }
+      }
+  }
+
+  private def pushWfNameAndInputsToMetadataService(workflowDescriptor: EngineWorkflowDescriptor): Future[Unit] = {
```

Comment:
In general I'd prefer that we model async behavior inside an actor via sub-actors as much as possible instead of mixing abstractions and using Futures. 

Particularly with these longer functions as then everything needs to be inspected every time there's a change to make sure nothing is captured